Add image export options

danfishgold opened the pull request:
Add the option to export the annotated code as an image (png), either as a download or pasted into the clipboard.

![Screen Shot 2021-08-05 at 19 21 19](https://user-images.githubusercontent.com/777023/128385516-268811b6-4106-4fa9-a776-023b36f7dc98.png).

This change should make it much easier to share annotations in a Miro board during code reading clubs.

The clipboard API doesn't allow jpeg images for some reason. I figured jpeg artifacts might be a problem with images that might be very big with very small text, so that constraint might be a good thing.

Also worth noting: [`Clipboard.write()`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/write) is not supported yet on Firefox, so the copy button is disabled there (via a check of `navigator.clipboard.write !== undefined`).

Both the download and copy options use [html2canvas](http://html2canvas.hertzen.com) to generate the image. There's a problem where it ignores the `mix-blend-mode` css rule, which is _very_ important for annotations, so I had to use [patch-package](https://www.npmjs.com/package/patch-package) to patch that in (I used the fix suggested in [this issue](https://github.com/niklasvh/html2canvas/issues/580#issuecomment-635450979)).

Exporting to pdf is probably very useful but it can't be done on the client side (other than printing to pdf). In order to solve this problem we'd have to do something server-side. This might be a good opportunity to combine this project and [CodeReadingClibs/pdf-maker](https://github.com/CodeReadingClubs/pdf-maker).

danfishgold committed:
Add copy and download buttons (partially working)

danfishgold committed:
Patch html2canvas to support mix-blend-mode

Thanks to [this comment](https://github.com/niklasvh/html2canvas/issues/580#issuecomment-635450979)
but with darken instead of multiply.

I used patch-package, which is really nice

danfishgold committed:
Add download functionality

danfishgold committed:
Extract export stuff to a different component

danfishgold committed:
Extract state management to a hook

danfishgold committed:
Add padding

danfishgold committed:
Remove exporting from the todo list!

katjam commented on README.md:
```diff
@@ -27,7 +27,6 @@ Netlify.
   near the top/bottom)
 - Footer with link to github and a "try your own code" button
 - Share links (needs hosting :((( )
-- Download/save/copy as image/pdf
```
:tada:

katjam commented on package.json:
```diff
@@ -4,11 +4,14 @@
     "build": "tsc && vite build",
-    "serve": "vite preview"
+    "serve": "vite preview",
+    "postinstall": "patch-package"
```
That's cool. I've not seen patches applied this way before.

danfishgold replied:
I saw it mentioned on twitter a couple of months ago and I was just waiting for the opportunity to use it 😄

katjam commented on src/hooks/useExport.ts:
```diff
@@ -0,0 +1,100 @@
+
+export type CopyState = 'idle' | 'preparing' | 'success' | 'failure'
+
+export type DownloadState = 'idle' | 'preparing' | 'failure'
```
I guess it makes sense not having success state here
vs
Needing a success state on copy so there is a noticeable side-effect.

danfishgold replied:
yeah, exactly! there's a one second timer that changes the button text to "copied"

katjam commented on src/components/Export.tsx:
```diff
@@ -0,0 +1,56 @@
+}
+
+function downloadButtonTitle(downloadState: DownloadState): string {
+  switch (downloadState) {
```
Loving the type safe pattern matches.

danfishgold replied:
you love to see it

katjam reviewed (approved):
This all looks pretty sensible to me. Thanks for the doc on the PR explaining your choices too.

danfishgold commented:
just one bug fix with regard to differences in `ClipboardItem` implementations between Chromium and WebKit and it's good to go

danfishgold committed:
Not sure what happened here...

This must have changed when I ran `npm install`

danfishgold committed:
Handle different ClipboardItem implementations across WebKit/Chromium

Chromium supports async stuff in user events, but not in ClipboardItem
and WebKit is the opposite.

(And Firefox doesn't support it at all)

Love to work with new web APIs!
